### PR TITLE
Initial FLE2 work

### PR DESCRIPTION
**Pull request by rozza**
Uses the latest [mongodb-crypt snapshot](https://oss.sonatype.org/content/repositories/snapshots/org/mongodb/mongodb-crypt/1.5.0-SNAPSHOT/).

Includes:

* AutoEncryptionOptions
* EncryptOptions
* Updated tests and added explicit prose tests.

### CODE REVIEW

File: driver-core/src/main/com/mongodb/internal/operation/CreateCollectionOperation.java
Comment (jyemin) [inline driver-core/src/main/com/mongodb/internal/operation/CreateCollectionOperation.java:402]: Nice

File: driver-core/src/main/com/mongodb/internal/operation/DropCollectionOperation.java
Comment (jyemin) [inline driver-core/src/main/com/mongodb/internal/operation/DropCollectionOperation.java:150]: Drop -> drop

File: driver-core/src/main/com/mongodb/internal/operation/DropCollectionOperation.java
Comment (jyemin) [inline driver-core/src/main/com/mongodb/internal/operation/DropCollectionOperation.java:161]: Is "ns not found" considered an error?

File: driver-core/src/main/com/mongodb/internal/operation/DropCollectionOperation.java
Comment (jyemin) [inline driver-core/src/main/com/mongodb/internal/operation/DropCollectionOperation.java:199]: I didn't realize we would have to call listCollections here.  Bummer.

**Review by jyemin — CHANGES_REQUESTED** [review]

File: driver-core/src/main/com/mongodb/AutoEncryptionSettings.java
Comment (rozza) [reply driver-core/src/main/com/mongodb/AutoEncryptionSettings.java:217]: None of the other builder setters define a default - so not sure why this should be different. 

Also this is complicated because as saying the default here (an empty map)  doesn't mean no fields are encrypted: 

> If a collection is not present on the encryptedFieldsMap a server-side collection encryptedFieldsMap may be used by the driver.

Comment (rozza) [reply driver-core/src/main/com/mongodb/AutoEncryptionSettings.java:217]: bypassAutoEncryption doesnt define the default - I think the reasoning is the name bypass is implicitly false.

File: driver-core/src/main/com/mongodb/AutoEncryptionSettings.java
Comment (rozza) [reply driver-core/src/main/com/mongodb/AutoEncryptionSettings.java:459]: See previous

File: driver-core/src/main/com/mongodb/AutoEncryptionSettings.java
Comment (rozza) [reply driver-core/src/main/com/mongodb/AutoEncryptionSettings.java]: Done

File: driver-core/src/main/com/mongodb/AutoEncryptionSettings.java
Comment (rozza) [reply driver-core/src/main/com/mongodb/AutoEncryptionSettings.java:201]: Done

File: driver-core/src/main/com/mongodb/AutoEncryptionSettings.java
Comment (rozza) [reply driver-core/src/main/com/mongodb/AutoEncryptionSettings.java]: Done

File: driver-core/src/main/com/mongodb/AutoEncryptionSettings.java
Comment (rozza) [reply driver-core/src/main/com/mongodb/AutoEncryptionSettings.java]: Done

File: driver-core/src/main/com/mongodb/client/model/CreateCollectionOptions.java
Comment (rozza) [reply driver-core/src/main/com/mongodb/client/model/CreateCollectionOptions.java:331]: Done

File: driver-core/src/main/com/mongodb/client/model/vault/EncryptOptions.java
Comment (rozza) [reply driver-core/src/main/com/mongodb/client/model/vault/EncryptOptions.java]: Done

File: driver-core/src/main/com/mongodb/client/model/vault/EncryptOptions.java
Comment (rozza) [reply driver-core/src/main/com/mongodb/client/model/vault/EncryptOptions.java]: :facepalm:

File: driver-sync/src/main/com/mongodb/client/internal/MongoDatabaseImpl.java
Comment (rozza) [reply driver-sync/src/main/com/mongodb/client/internal/MongoDatabaseImpl.java:122]: Good catch

File: driver-sync/src/main/com/mongodb/client/internal/CommandMarker.java
Comment (rozza) [reply driver-sync/src/main/com/mongodb/client/internal/CommandMarker.java:40]: Good Catch

File: driver-core/src/main/com/mongodb/internal/operation/DropCollectionOperation.java
Comment (rozza) [reply driver-core/src/main/com/mongodb/internal/operation/DropCollectionOperation.java]: Done.

File: driver-core/src/main/com/mongodb/internal/operation/DropCollectionOperation.java
Comment (rozza) [reply driver-core/src/main/com/mongodb/internal/operation/DropCollectionOperation.java:161]: Not sure - this logic is due to change in DRIVERS-2326 so have asked there: https://github.com/mongodb/specifications/pull/1215

File: driver-core/src/main/com/mongodb/AutoEncryptionSettings.java
Comment (jyemin) [reply driver-core/src/main/com/mongodb/AutoEncryptionSettings.java:201]: I don't see that this was changed.

Comment (rozza) [reply driver-core/src/main/com/mongodb/AutoEncryptionSettings.java]: Should be there now

File: driver-sync/src/main/com/mongodb/client/internal/MongoCollectionImpl.java
Comment (rozza) [reply driver-sync/src/main/com/mongodb/client/internal/MongoCollectionImpl.java]: Done

**Comment by jyemin** [comment]
https://spruce.mongodb.com/task/mongo_java_driver_tests_jdk_secure__version~4.2_os~linux_topology~replicaset_auth~auth_ssl~ssl_jdk~jdk8_test_patch_5f0cf4b43be0f9e25aa65566238d7e6907f870db_62838a130ae60667d2b881d9_22_05_17_11_42_13/tests?execution=0&sortBy=STATUS&sortDir=ASC looks like a new failure

**Comment by rozza** [comment]
> https://spruce.mongodb.com/task/mongo_java_driver_tests_jdk_secure__version~4.2_os~linux_topology~replicaset_auth~auth_ssl~ssl_jdk~jdk8_test_patch_5f0cf4b43be0f9e25aa65566238d7e6907f870db_62838a130ae60667d2b881d9_22_05_17_11_42_13/tests?execution=0&sortBy=STATUS&sortDir=ASC looks like a new failure

@jyemin Should be fixed now, however, the encryption tests are also failing on 6.0 builds - any idea what version is installed?  Tested locally using 6.0.0-rc5

**Comment by jyemin** [comment]
[This](https://github.com/mongodb-labs/drivers-evergreen-tools/blob/master/.evergreen/download-mongodb.sh#L57) indicates rc4

File: driver-core/src/main/com/mongodb/client/model/CreateCollectionOptions.java
Comment (markbenvenuto) [reply driver-core/src/main/com/mongodb/client/model/CreateCollectionOptions.java:360]: No. There are no secrets here. The key ids are identifiers, not secrets.

File: driver-core/src/main/com/mongodb/AutoEncryptionSettings.java
Comment (markbenvenuto) [reply driver-core/src/main/com/mongodb/AutoEncryptionSettings.java:219]: They are basically the same. I am not sure why Kevin introduced `bypassQueryAnalysis`.

File: driver-core/src/main/com/mongodb/AutoEncryptionSettings.java
Comment (markbenvenuto) [reply driver-core/src/main/com/mongodb/AutoEncryptionSettings.java:460]: The statement is correct. Underneath, libmongocrypt has to append the schema to the query/CRUD command so that the server can finish the indexing/query.

File: driver-core/src/main/com/mongodb/client/model/vault/EncryptOptions.java
Comment (markbenvenuto) [reply driver-core/src/main/com/mongodb/client/model/vault/EncryptOptions.java:124]: Range: 0 to Long.MAX_VALUE. Negative numbers are rejected by the server.

File: driver-core/src/main/com/mongodb/internal/operation/DropCollectionOperation.java
Comment (markbenvenuto) [inline driver-core/src/main/com/mongodb/internal/operation/DropCollectionOperation.java:180]: Per https://github.com/mongodb/specifications/pull/1215/, do the drop of the user collection last. We warn users in the log file to clean up the esc/ecc/ecoc collections if they drop the user collection first.

File: driver-core/src/main/com/mongodb/internal/operation/CreateCollectionOperation.java
Comment (markbenvenuto) [inline driver-core/src/main/com/mongodb/internal/operation/CreateCollectionOperation.java:466]: https://github.com/mongodb/specifications/pull/1215/ makes this clustered collection.

**Review by markbenvenuto — APPROVED** [review]

File: driver-core/src/main/com/mongodb/AutoEncryptionSettings.java
Comment (jyemin) [reply driver-core/src/main/com/mongodb/AutoEncryptionSettings.java:219]: @kevinAlbs is it worth discussing this?

File: driver-core/src/main/com/mongodb/client/model/vault/EncryptOptions.java
Comment (jyemin) [reply driver-core/src/main/com/mongodb/client/model/vault/EncryptOptions.java:124]: Let's add that to the param description then.

**Review by jyemin — CHANGES_REQUESTED** [review]

File: driver-core/src/main/com/mongodb/internal/operation/DropCollectionOperation.java
Comment (rozza) [reply driver-core/src/main/com/mongodb/internal/operation/DropCollectionOperation.java:180]: Ack will do in JAVA-4609 - currently blocked until merge.

File: driver-core/src/main/com/mongodb/internal/operation/CreateCollectionOperation.java
Comment (rozza) [reply driver-core/src/main/com/mongodb/internal/operation/CreateCollectionOperation.java:466]: Ack will do in [JAVA-4609](https://jira.mongodb.org/browse/JAVA-4609) - currently blocked until merge.

File: driver-core/src/main/com/mongodb/client/model/vault/EncryptOptions.java
Comment (rozza) [reply driver-core/src/main/com/mongodb/client/model/vault/EncryptOptions.java:124]: Done

File: driver-core/src/main/com/mongodb/AutoEncryptionSettings.java
Comment (rozza) [reply driver-core/src/main/com/mongodb/AutoEncryptionSettings.java:219]: Perhaps a question better suited for #drivers-fle or a spec ticket.

File: driver-core/src/main/com/mongodb/internal/operation/DropCollectionOperation.java
Comment (rozza) [reply driver-core/src/main/com/mongodb/internal/operation/DropCollectionOperation.java:161]: Confirmed it wont be considered an error.

**Comment by rozza** [comment]
WIll add [JAVA-4609](https://jira.mongodb.org/browse/JAVA-4609) shortly.

File: driver-core/src/main/com/mongodb/AutoEncryptionSettings.java
Comment (kevinAlbs) [reply driver-core/src/main/com/mongodb/AutoEncryptionSettings.java:219]: Please see [Q3 of the scope](https://docs.google.com/document/d/1duvPDqbFFEvSjTWX_8Vp-u_IjGZGdV3ZyrzCjrjbn2E/edit#heading=h.he9g3svb42o7).

`bypassAutoEncryption` permits the driver to not pass the command through libmongocrypt. The driver can bypass the reduced batch splitting.

`bypassQueryAnalysis` requires the driver to pass the command through libmongocrypt. It is useful for FLE 2 to append additional data to outgoing commands.

Comment (jyemin) [reply driver-core/src/main/com/mongodb/AutoEncryptionSettings.java:219]: It's also in the spec: https://github.com/mongodb/specifications/blob/master/source/client-side-encryption/client-side-encryption.rst#why-is-bypassqueryanalysis-needed

Resolving.

**Review by jyemin — APPROVED** [review]
LGTM!

**Comment by rozza** [comment]
@jyemin please review the last commit - added the clustered index and changed the drop order

**Review by jyemin — APPROVED** [review]
LGTM again

**Review by markbenvenuto — APPROVED** [review]

**Comment by rozza** [comment]
Manually fixed up / squashed rebased and merged

**Pull request closed without merging**
